### ISSUE TITLE
matching behavior, strings with slashes

### ISSUE THREAD

**Issue by evan-phelps** [OHDSI/Usagi#14]
I don't know if this might be a Lucene issue or quirk, or if I might have a misunderstanding about what I should expect from Usagi's matching behavior.

The following source description fails to match on an RxNorm concept who's description is identical:

Sodium Chloride 0.154 MEQ/ML Injectable Solution

The issue is not simply that the score is too low; instead, the RxNorm concept with the identical description is not listed in the Results panel even though it exists.  More specifically, the behavior occurs as soon as the slash in "MEQ/ML" is introduced to the query.  For example, querying "Sodium Chloride 0.154 MEQ"  matches with a score of 0.81 to the above description, but "Sodium Chloride 0.154 MEQ/" leads to the above description disappearing from the Results panel altogether.  The result remains absent from the Results panel as more characters, matching or otherwise, are added to the query string.

**Comment by schuemie** [comment]
I can reproduce the problem, but I don't think it has to do with the slashes. Instead, consider the 'best' match I find when using 'Sodium Chloride 0.154 MEQ/ML Injectable Solution' as the search string:

Potassium Chloride 0.4 MEQ/ML / Sodium Chloride 0.154 MEQ/ML Injectable Solution

This not only has everything in the search string, it has some parts twice! Clearly, this must be a better match ;-) I'm also guessing the parts that match twice (e.g. 'MEQ/ML') are pretty rare in the vocab in general, and therefore get a high weight. These other matches simply crowd out the concept you're looking for.

I do not know how to modify Lucene's behavior in this case. However, I would like to point out that in general, for drug mapping, we do not use Usagi to directly map from drug code to drug code. Instead, we typically use the structured nature of the codes and the vocabulary. We create an ingredient-to-ingredient map, a unit-to-unit map, and a drug form-to-drug form map (maybe using Usagi), and then use SQL queries to find the right standard concepts (usually RxNorm, but there are non-RxNorm drug concepts as well) to map to based on the combination of ingredient-strength-form. Please see the drug_strength table in the vocabulary on how this information is available.

**Comment by evan-phelps** [comment]
Thank you for your response and extra information about how drug codes are typically treated. 

I can justify a lower match score based on the TF-IDF logic, but I wasn't able to rationalize why the exact match completely disappeared from the scored list of results.  That's still mysterious to me, but I understand that, one way or another, it's a behavior of Lucene.  I might dig into that, even though my example was somewhat contrived -- i.e., I had already parsed, normalized, matched, and resolved to RxNorm before passing the codes through Usagi.

My original source codes and descriptions did not seem amenable to the Usagi treatment, which maybe relates to your guidance on the typical method of drug mapping.  I'll need to think about what you described after reviewing documentation carefully, because I don't have a "tight" understanding of you what you are suggesting.

In the meantime, please entertain/critique this (or feel free to send me on my way to RTM):  Outside of Usagi, I already programmatically resolved my non-standard drug codes to OMOP concept IDs of maximal information content, so I can easily populate the concept mappings that connect my non-standard source codes to the standardized vocabularies.  But is this incompatible with the typical approach that you described?  I'm afraid my thinking might have been the opposite of typical in that I first mapped to maximum-content standardized codes (in this case, RxNorm and corresponding OMOP concepts) allowing for easy decomposition into dose form, single-ingredient components, and strengths independent of OMOP.

**Comment by schuemie** [comment]
The reason the 'TRUE' match disappears is that I cut off the list of results. Because we're using n-gram matching, pretty much every search returns many hits, and it would slow the UI down tremendously to show them all.

To try and explain what I mean in more detail, take for example some work I did in trying to map Japanese drug codes. Here, the source data contained, for each drug code, separate fields for the generic name (usually the ingredient), the brand name, the strength, and the form, as well as ATC codes that were linked to the drug codes.

Step 1 was linking all ingredients to ingredients in the Vocab. For that purpose, the ATC codes were extremely helpful because (a) most ATC codes are 1-on-1 linked to an ingredient and (b) ATC is also in the Vocab. For those where the ATC linked to more than 1 ingredient, I used string matching to try and find the right match. 

Step 2 was linking all drug strength units (e.g. mapping 'milligram' to its appropriate concept ID in the Vocab).

Step 3 was linking all drug forms (e.g. mapping 'oral tablet' to its appropriate concept ID in the Vocab). 

I can now use this **structured** information to find the best links for my drug codes. Using the drug_strength table, I can list all drugs that contain a certain ingredient. I can find which ones match the right strength and form. If I do not find an exact map, which happens very often because a certain strength used in one part of the world might not be used in the US, I can choose to have a less exact match, for example by just mapping to the ingredient instead.

Hope this make it a bit clearer.
